[PAC][Driver] Support ptrauth flags only on ARM64 Darwin or with pauthtest ABI

kovdan01 opened the pull request:
Most ptrauth flags are ABI-affecting, so usually we do not want them to be
exposed to end users. Allow them only in the following cases:

- ARM64 Darwin (under certain conditions, some ptrauth driver flags are
  intended to be used in this case);
- pauthtest ABI (it's intended to be used for experimenting with signing schema
  and the signing schema is explicitly encoded in the pauth elf marking).

Leave `-faarch64-jump-table-hardening` available for all AArch64 targets
since it's not ABI-affecting.

kovdan01 commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/113152?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#113152** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/113152?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#113151** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/113151?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#113150** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/113150?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#113149** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/113149?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @kovdan01 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

kovdan01 commented:
Would be glad to see everyone's feedback on the changes.

kovdan01 commented:
Would be glad to see everyone's feedback on the changes.

kovdan01 commented:
@MaskRay As discussed in https://github.com/llvm/llvm-project/pull/96160#issuecomment-2468793288, this drops support for ptrauth driver flags, but with an exception for apple targets - to the best of my knowledge, they have use cases for them. Also tagging @ahmedbougacha 

Note: as stated in PR description, the jump table hardening flag (#113149) is left intact since it's not ABI-affecting.

asl commented:
I think we'd need to keep driver flags for `pauthtest` ABI as the intention is to enable easier bringup of pointer authentication ABI on a new target:
  - `pauthest` is intended to be used for experimenting with signing schema w/o changing compiler
  - the signing schema is explicitly encoded in the pauth elf marking

Though certainly for other *named* ABIs (say, platform dependent) they should be disabled by default and it is up to platform to decide on signing schema.

kovdan01 commented:
> I think we'd need to keep driver flags for `pauthtest` ABI as the intention is to enable easier bringup of pointer authentication ABI on a new target

@asl Thanks for suggestion, addressed in 515574c415b7d9814d7942c7609095dcb3e7916a, and updated the PR description correspondingly.

MaskRay commented on clang/lib/Driver/ToolChains/Linux.cpp:
```diff
@@ -543,12 +543,59 @@ static void handlePAuthABI(const Driver &D, const ArgList &DriverArgs,
     CC1Args.push_back("-faarch64-jump-table-hardening");
 }
 
+static void addPointerAuthFlags(const llvm::opt::ArgList &DriverArgs,
```
There is a lot of duplication with Darwin.cpp.  Should be moved to CommonArgs.cpp

kovdan01 replied:
Thanks for suggestion, moved common code to `ToolChain::addPointerAuthFlags`, see 3afcc78d48cb49fb434478490c3ee79e7525219c

MaskRay commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -1808,34 +1808,6 @@ void Clang::AddAArch64TargetArgs(const ArgList &Args,
 
   AddUnalignedAccessWarning(CmdArgs);
 
-  Args.addOptInFlag(CmdArgs, options::OPT_fptrauth_intrinsics,
```
https://maskray.me/blog/2023-08-25-clang-wunused-command-line-argument#target-specific-options

You can make these options TargetSpecific and only handle them for Linux and Darwin. Then these options will lead to errors for other platforms.

kovdan01 replied:
> You can make these options TargetSpecific and only handle them for Linux and Darwin.

@MaskRay It looks like they are already declared as TargetSpecific - see https://github.com/llvm/llvm-project/blob/41cde46/clang/include/clang/Driver/Options.td#L4337-L4357. Please let me know if I'm missing smth and you meant smth else.

MaskRay replied:
I'm afraid that you missed the point. Keep the code in Clang.cpp but do the following

```
// Clang.cpp
if (Linux pauth abi or isDarwin) {
  claim these pauth options
}
```

asl replied:
@MaskRay This would explicitly require adding each platform that would want to add pauth here. I don't think this is something desired (do we really want to add further, say, OpenBSD, FreeBSD and other platforms in the condition?)

MaskRay replied:
Yes for such straightforward CC1 forwarding options. The alternative requires a new function in CommonArgs.cpp and changes to various ToolChains/XX.cpp files, which is worse.

kovdan01 replied:
@MaskRay Applied your suggestion in 346d1ef9d0d638ddda2b48a9c629f24a4385df02

> I'm afraid that you missed the point. Keep the code in Clang.cpp but do the following
> 
> ```
> // Clang.cpp
> if (Linux pauth abi or isDarwin) {
>   claim these pauth options
> }
> ```

MaskRay reviewed (changes requested):
.

kovdan01 commented:
@MaskRay Previous comments should now be addressed, would be glad to see your feedback on the new version of the PR

MaskRay commented on clang/test/Driver/aarch64-ptrauth.c:
```diff
@@ -15,50 +16,85 @@
 // RUN:   -fno-ptrauth-type-info-vtable-pointer-discrimination -fno-ptrauth-indirect-gotos \
 // RUN:   -fno-ptrauth-init-fini -fno-ptrauth-init-fini-address-discrimination \
-// RUN:   -fno-aarch64-jump-table-hardening %s 2>&1 | FileCheck %s --check-prefix=PAUTHABI2
+// RUN:   -fno-aarch64-jump-table-hardening %s 2>&1 | FileCheck %s --check-prefix=PAUTHTEST2
```
It's not necessary to change PAUTHABI test prefixes to PAUTHTEST. Unnecessary churn.

kovdan01 replied:
Changed back to PAUTHABI, thanks

kovdan01 committed:
[PAC][Driver] Support ptrauth flags only on ARM64 Darwin or with pauthtest ABI

Most ptrauth flags are ABI-affecting, so usually we do not want them to be
exposed to end users. Allow them only in the following cases:

- ARM64 Darwin (under certain conditions, some ptrauth driver flags are
  intended to be used in this case);
- pauthtest ABI (it's intended to be used for experimenting with signing schema
  and the signing schema is explicitly encoded in the pauth elf marking).

Leave `-faarch64-jump-table-hardening` available for all AArch64 targets
since it's not ABI-affecting.
